enhancements/cluster-logging/forward-to-loki.md

```diff
@@ -0,0 +1,221 @@
+  - Must not allow distinct nodes with separate clocks to contribute to a single stream.
+
+All logging meta-data will still be included in log records.
+Labels do not need to _identify the source of logs_, only to _partition the search space_.
+At query time labels reduce the search space, then Loki uses log content to complete the search.
+
+**Note**: Loki [label names][] must match the regex `[a-zA-Z_:][a-zA-Z0-9_:]*`.
+Meta-data names are translated to Loki labels by replacing illegal characters with '_'.
+
+Nested JSON objects are referenced by converting their JSON path to a label names.
+For example, if `kubernetes.namespace_name` is a label, then this JSON log record:
+
+``` json
+{"kubernetes.namespace_name":"foo","kubernetes":{"labels":{"foo":"bar"}}
+```
+would match this LogQL filter: 
+
+``` logql
+{ kubernetes_namespace_name="foo" } | json | kubernetes_labels_foo == "bar"
+```
+### Default Loki Labels
+
+The proposed default label set is:
+
+- `log_type`: one of "application", "infrastructure", "audit"
+- `cluster`: for muti-cluster deployments (see Open Questions).
+- `kubernetes.host`: aka node, ensures ordered streams.
+- `kubernetes.namespace_name`: likely to be most common known factor in a search.
+- `kubernetes.container_name`: low cardinality, usually well-known
+
+`kubernetes.pod_name` is _not_ included. It is potentially high cardinality.
+`kubernetes.namespace_name` can be used to constrain the search for a named pod.
+
+The API below allows the user to add labels that are relevant for their application.
+
+### Proposed API
+
+Add a new `loki` output type to the `ClusterLogForwarder` API:
+
+``` yaml
+- name: myLokiOutput
+  type: loki
+  url: ...
+  secret: ...
+```
+
+`url` and `secret` have the usual meaning with regard to TLS and certificates.
+The `secret` may also contain `username` and `password` fields for Loki.
+
+The following optional fields are Loki-specific:
+
+- `labelKeys`: ([]string, optional)  A list of meta-data keys to add to the label set.\
+  Keys are translated to Loki labels: `kubernetes.labels.foo` => `kubernetes_labels_foo`,
+  both the dotted and underscored forms are allowed in the list.
+
+_**TODO**: Link to data model documentation for available meta-data keys._
+
+_**TODO**: Review Loki client configuration for any other fields._
+
+### User Stories
+
+#### See all logs from a namespace
+
+``` logql
+{ kubernetes_namespace_name="mynamespace"}
+```
+
+#### See logs from a specific container in a named Pod
+
+``` logql
+{ kubernetes_namespace_name="mynamespace", kubernetes_container_name="mycontainer" } |= kubernetes_pod_name == "mypod"
+```
+
+#### See logs from a labelled application
+
+Using the default configuration:
+
+``` logql
+{ } |= kubernetes_labels_app == "myapp""
+```
+
+By adding `labelKeys: [ kubernetes.labels.app ]` to the Loki output configuration this
+becomes a faster label query:
+
+``` logql
+{ kubernetes_labels_app="myapp" }
+```
```

Comment:
I'm not seeing how assigning labels to a Loki output assigns them to a particular application.